### PR TITLE
Sync content to Azure DevOps

### PR DESCRIPTION
**Pull request by live1206**
[AB#454852](https://dev.azure.com/ceapex/Engineering/_workitems/edit/454852)

### CODE REVIEW

File: sync-github-vsts.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# variables:
```
Comment (yufeih) [inline sync-github-vsts.yml:1]: `vsts` is an old term, it is nowdays called `azure-repos`.

File: sync-github-vsts.yml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  # branches: defined on azure portal, separated by '|'
+  # vstsUrl: defined on azure portal
+
+trigger:
+  batch: true
+  branches:
+    include:
```
Comment (yufeih) [inline sync-github-vsts.yml:8]: Now that we have many pipelines (and potentially more after OneBranch), consider move all azure pipeline yaml to the `.pipelines` folder to match OneBranch convention.

File: sync-github-vsts.yml [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# variables:
+  # branches: defined on azure portal, separated by '|'
```
Comment (yufeih) [inline sync-github-vsts.yml:2]: This is comment out of date? I see no `|`

File: sync-github-vsts.yml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    - v3
+    - v3-release
+    - dev
+    - main
+
+variables:
+  NugetSecurityAnalysisWarningLevel: none
+  
+jobs:
+- job: 'GitHub_Sync'
```
Comment (yufeih) [inline sync-github-vsts.yml:18]:
```suggestion
- job: 'SyncGitHubToAzureRepos'
```

File: sync-github-vsts.yml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  pool: 
+    vmImage: 'windows-latest'
+  steps:
+  - task: Powershell@2
+    displayName: sync docfx v3 from GitHub
```
Comment (yufeih) [inline sync-github-vsts.yml:23]: Not just v3?

File: sync-github-vsts.yml [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+          git checkout -b $branch
+          $B64_PAT=[System.Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes("user:$(System.AccessToken)"))
+          git -c http.extraHeader="Authorization: Basic $B64_PAT" push $(vstsUrl) $branch -uf
+        }
+
+        $folderPath = "temp"
+        mkdir $folderPath
+        pushd $folderPath
+        git init
+        foreach($branch in "$(branches)".Split(',').trim()) {
```
Comment (yufeih) [inline sync-github-vsts.yml:41]: Suppose this is hook to commit trigger, we only need to sync the current branch right?

File: sync-github-vsts.yml (hunk 2 above)
Comment (live1206) [reply sync-github-vsts.yml:8]: Prefer to do it when invloving OneBranch

File: sync-github-vsts.yml [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      script: |
+        Function Sync([string]$branch) {
+          git fetch $(Build.Repository.Uri) $branch
+          git checkout FETCH_HEAD
+          git checkout -b $branch
+          $B64_PAT=[System.Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes("user:$(System.AccessToken)"))
+          git -c http.extraHeader="Authorization: Basic $B64_PAT" push $(ADOUrl) $branch -uf
+        }
+
+        $folderPath = "temp"
```
Comment (yufeih) [inline sync-github-vsts.yml:36]: Do we still need this `temp` folder now that there is only a single branch?

**Comment by mingwli** [comment]
> 🚗🌳💥🤕🚑[azure-docs-rest-apis](https://github.com/Azure/azure-docs-rest-apis)(00:00:00)

new service account doesn't have access to this repo yet, please ignore the crash

File: sync-github-vsts.yml [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    displayName: sync from GitHub to Azure DevOps
+    env:
+      SYSTEM_ACCESSTOKEN: $(System.AccessToken)
+    inputs:
+      targetType: inline
+      script: |
+        Function Sync([string]$branch) {
+          git fetch $(Build.Repository.Uri) $branch
+          git checkout FETCH_HEAD
+          git checkout -b $branch
```
Comment (yufeih) [inline sync-github-vsts.yml:31]: should the repository already been implicitly checked out by azure pipeline? All we need is a `git push`?

File: sync-github-vsts.yml (hunk 2 above)
Comment (yufeih) [reply sync-github-vsts.yml:7]: Consider change the file name to `sync-github-to-azure-repos.yml`

File: sync-github-vsts.yml (hunk 8 above)
Comment (live1206) [reply sync-github-vsts.yml:31]: Not really, you need to checkout the specific branch firstly

File: .github/actions/spelling/expect.txt [hunk 9]
```diff
@@ -956,3 +956,5 @@ yyyy
 zoneinfo
 zonepivotgroups
 zzz
+popd
```
Comment (yufeih) [inline .github/actions/spelling/expect.txt:959]: revert?

File: sync-github-ado.yml [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    env:
+      SYSTEM_ACCESSTOKEN: $(System.AccessToken)
+    inputs:
+      targetType: inline
+      script: |
+        $branch = "$(Build.SourceBranch)"
+        git init
+        git fetch $(Build.Repository.Uri) $branch
+        git checkout FETCH_HEAD
+        git checkout -b $branch
```
Comment (yufeih) [inline sync-github-ado.yml:32]: > Not really, you need to checkout the specific branch firstly

We could use `git push https://xyz HEAD:$branch`

File: sync-github-ado.yml [hunk 11]
```diff
@@ -0,0 +1,34 @@
+# variables:
+  # ADOUrl: defined on azure portal
```
Comment (yufeih) [inline sync-github-ado.yml:2]: Consider define the azure repo URL directly here in code over a config in azure pipelines. Code is the source of truth for everything.

File: sync-github-ado.yml [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+variables:
+  NugetSecurityAnalysisWarningLevel: none
+  
+jobs:
+- job: 'Sync_GitHub_To_AzureDevOps'
+  pool: 
+    vmImage: 'windows-latest'
+  steps:
+  - task: Powershell@2
+    displayName: sync from GitHub to Azure DevOps
```
Comment (yufeih) [inline sync-github-ado.yml:22]:
```suggestion
    displayName: Sync from GitHub to Azure DevOps
```

File: sync-github-ado.yml (hunk 10 above)
Comment (live1206) [reply sync-github-ado.yml:32]: image
Not able to push without upstream branch.

File: sync-github-ado.yml (hunk 11 above)
Comment (live1206) [reply sync-github-ado.yml:2]: Do not think it's good to expose the internal Azure DevOps repository Url in the public GitHub repo

**Review by herohua — CHANGES_REQUESTED** [review]
Please hold off this PR until we get confirmation whether we can have private data exposed in public repo.

**Comment by live1206** [comment]
> Please hold off this PR until we get confirmation whether we can have private data exposed in public repo.

Offline discussed with @yufeih , we can still proceed with this approach if we put OneBranch YAML within ADO repository instead

File: .github/actions/spelling/expect.txt [hunk 13]
```diff
@@ -955,4 +955,4 @@ yyy
 yyyy
 zoneinfo
 zonepivotgroups
-zzz
+zzz
```
Comment (yufeih) [inline .github/actions/spelling/expect.txt:958]: There is still diff in this file

Comment (live1206) [reply .github/actions/spelling/expect.txt]: reset

**Pull request closed without merging**